### PR TITLE
Rework db keys (YRCORE-3958)

### PR DESCRIPTION
**Pull request by ogrubes**
Still TODO:
1. ~~Please take a look at the non-timeline "TODOs" near the end of this review. I should either fix, ticket, or remove each of these before submitting this branch.~~
2. ~~`go test $(glide nv)` currently failing at transicator/test [build failed] https://pb.yelpcorp.com/260123~~
3. ~~I should either rework or replace `testIndexKeyCompare` before submitting this branch. Byte-array comparison logic is currently/unnecessarily written out in this test.~~ 
**^^ NOTE: I'm leaving in manual byte-comparison logic after all**
4. ~~remove sort.Sort() from GetMultiEntries and rework readOneRange to accommodate.~~

## Included in this review:
- reorder the db key from `scope_len|scope|lsn|index` to `lsn|index|scope`
- remove golang key comparator (entryComparator)
- remove test-only `GetEntries` in favor of `GetMultiEntries` (which is actually used in code)
- rework `GetMultiEntries` to read more efficiently for a list of scopes, and eliminate Sort() call
- rework storage_test structure
- more key conversion tests in convert_test

## Not included in this review:
- `timeline` will be added to the key in a separate ticket.

### CODE REVIEW

**Comment by kriswehner** [comment]
On your above comments:
2. go test $(glide nv) will always fail on the test/ package, as it expects docker stuff. Don't fix it.

**Comment by ogrubes** [comment]
Check out the 3rd commit, now with readOneRange reborn

File: storage/convert_test.go [hunk 1]
```diff
@@ -2,94 +2,172 @@ package storage
...
 )
 
 var _ = Describe("Conversion", func() {
-	It("Identity Key Conversion", func() {
-		key := lsnAndOffsetToKey("foo", 3, 5)
-		scope, lsn, offset := keyToLsnAndOffset(key)
-		Expect(scope).Should(BeEquivalentTo("foo"))
-		Expect(lsn).Should(BeEquivalentTo(3))
-		Expect(offset).Should(BeEquivalentTo(5))
+	It("Index Key Compare Quick Check", func() {
```
Comment (kriswehner) [inline storage/convert_test.go]: Can we drop a "should" in here and make these read well?

File: storage/storage.go [hunk 2]
```diff
@@ -319,31 +300,26 @@ func (s *DB) readOneRange(scope string, lsn uint64,
...
 		iterKey := it.Key().Data()
-		iterScope, iterLSN, iterIndex := keyToLsnAndOffset(iterKey)
-		if iterScope != scope { // TODO: string comparison?
-			// NOTE: this is inefficient.
-			// See method's use within GetMultiEntries, and consider
-			// having this func take in multiple scopes instead.
-			// This may mean adding scope to readResult struct,
-			// or using some other way to associate results:scope.
+		iterScope, _, _ := keyToLsnAndOffset(iterKey)
+		if !scopesMap[iterScope] && len(scopesMap) > 0 {
```
Comment (kriswehner) [inline storage/storage.go]: This isn't the way I typically see go maps used. I think it's like:

if _, ok = scopesMap[iterScope]; !ok {
   // Stuff to do if key is missing from the map
}

File: storage/storage.go [hunk 3]
```diff
@@ -240,16 +232,11 @@ func (s *DB) GetMultiEntries(scopes []string, metadataKeys []string,
 	ropts.SetSnapshot(snap)
 	defer ropts.Destroy()
 
-	// Read range for each scope
-	// TODO: should we store scope information in readResult?
-	// TODO: should we at least store scope information in returned results here?
+	// Read range for the list of scopes
 	var results readResults
```
Comment (kriswehner) [inline storage/storage.go]: I would vote: Kill this type alias and just declare this as "var results [][]byte"

File: storage/storage.go [hunk 4]
```diff
@@ -40,13 +38,7 @@ type DB struct {
...
 }
 
-// TODO: YRCORE-4003 add timeline here
-type readResult struct {
-	lsn   uint64
-	index uint32
-	data  []byte
-}
-type readResults []readResult
+type readResults [][]byte
```
Comment (kriswehner) [inline storage/storage.go]: I think this can die.

File: storage/storage.go (hunk 3 above)
Comment (ogrubes) [inline storage/storage.go]: as chatted, going to leave this (as a naming convention) to make it clear that the [][]bytes we are using here are the "readResults" we are interested in.

There are a lot of []bytes going around, so this was intended to distinguish the output.
Having this alias will also make it obvious how to change the returned data type / temp stored values if we decide to change the logic in the future.

File: storage/storage.go [hunk 5]
```diff
@@ -27,49 +19,6 @@ const (
...
-	return scopeCmp
-}
-
-/*
-Name is part of the comparator interface.
-*/
-func (c entryCmp) Name() string {
-	return ComparatorName
-}
-
```
Comment (kmorrison) [inline storage/storage.go]: get ye gone

File: storage/storage.go [hunk 6]
```diff
@@ -380,44 +285,41 @@ func (s *DB) PurgeEntries(filter func([]byte) bool) (purgeCount uint64, err erro
...
+	}
 
 	for ; it.Valid() && len(results) < limit; it.Next() {
 		iterKey := it.Key().Data()
-		if entryComparator.Compare(iterKey, endKeyBuf) > 0 {
-			// Reached the end of our range
-			//fmt.Printf("Stopped due to end of range, iterKey is %s, endKey is %s\n", hex.Dump(iterKey), hex.Dump(endKeyBuf))
-			break
+		iterScope, _, _ := keyToLsnAndOffset(iterKey)
+		if _, ok := scopesMap[iterScope]; !ok && len(scopesMap) > 0 {
```
Comment (kmorrison) [inline storage/storage.go]: This bears a comment. Rawr.

File: storage/storage.go [hunk 7]
```diff
@@ -380,44 +285,41 @@ func (s *DB) PurgeEntries(filter func([]byte) bool) (purgeCount uint64, err erro
...
+	startKeyBuf := lsnAndOffsetToKey("", lsn, index)
 	it.Seek(startKeyBuf)
 
 	var results readResults
+	scopesMap := make(map[string]bool, len(scopes))
+	for _, scope := range scopes {
+		scopesMap[scope] = true
+	}
 
 	for ; it.Valid() && len(results) < limit; it.Next() {
```
Comment (kmorrison) [inline storage/storage.go]: Let's create a ticket for load-testing reading 100 milliweeks of data for v1.

Comment (kriswehner) [inline storage/storage.go]: Just to close the loop: I'm solving this with the load testing tool.

Comment (ogrubes) [inline storage/storage.go]: 👍  added a comment to YRCORE-4000

**Comment by ogrubes** [comment]
Clean test run: https://pb.yelpcorp.com/260801

Should be ready for merge!

File: storage/storage.go (hunk 6 above)
Comment (ogrubes) [inline storage/storage.go]: 👍

**Pull request merged**